### PR TITLE
16bit typeid

### PR DESCRIPTION
**Pull request by smessmer**
Only use 16bit instead of 64bit for a type id.

This reduces the space needed for dispatch keys from 10 to 4 bytes per tensor, improving cache efficiency, alignment and comparison time when looking up kernels from the dispatch hash table.

### CODE REVIEW

File: caffe2/core/typeid.h [hunk 1]
```diff
@@ -162,6 +162,8 @@ class TypeMeta {
   // in .cpp to manage dependencies
   static void _ThrowRuntimeTypeLogicError(const std::string& msg);
 
+  static CaffeTypeId _createTypeId();
```
Comment (Yangqing) [inline caffe2/core/typeid.h]: Should this be probably a static member function of CaffeTypeId, instead of TypeMeta?

**Review by Yangqing — COMMENTED** [review]
Neat :)

**Comment by ezyang** [comment]
What's the fbcode test plan?

**Comment by smessmer** [comment]
@ezyang Do you think this is at risk of breaking anything? I was hoping our unit test coverage is good enough to catch likely errors here. What kind of tests do you propose?

File: caffe2/core/typeid.h (hunk 1 above)
Comment (smessmer) [reply caffe2/core/typeid.h]: Good idea, this allowed me to remove the constructors from CaffeTypeId, leaving this the only way to create instances.

**Comment by ezyang** [comment]
I'm thinking mostly perf-wise, since this is a performance oriented change.

**Comment by smessmer** [comment]
I don't expect immediate perf benefits since the type id size isn't on a critical path yet. It will be with the new dispatcher.

**Comment by Yangqing** [comment]
Let's make sure that there is a microbenchmark in place to test this - whether this is now or with the new dispatcher is fine with me, and I think it is reasonable expectation that this improves perf - but it is a little bit nonconventional to claim a (potential) performance win without benchmarking code.

**Review by ezyang — APPROVED** [review]
Please pay attention when the fbcode sync happens; this is the kind of thing that is higher risk for fbcode use cases ;)

**Pull request merged**